### PR TITLE
Progress Report 2017 Fourth Quarter

### LINKED ISSUE #65: Progress Report 2017 October
Double the progress report, double the ~~torture~~ FUN!

RFC will be open from 2017-11-25 08:00-0400 to 2017-12-10 23:59-0400.
Only PRs that were *merged* between 2017-10-01 and 2017-10-31.

### LINKED ISSUE #66: Progress Report 2017 November
Double the progress report, double the ~~torture~~ FUN!

RFC will be open from 2017-11-25 08:00-0400 to 2017-12-10 23:59-0400.
Only PRs that were *merged* between 2017-11-01 and 2017-11-30.

### LINKED ISSUE #67: Progress Report 2017 December
Unlike #65 and #66, because December is still ongoing, this is a living breathing document, that should have PRs added to it as they are merged.

RFC will be open from 2017-12-11 09:15-0400 to 2018-01-02 09:15-0400.
Again, make sure that the PR has *already* been merged to `master` before adding a comment here.

Wow, can't believe the year's almost over.

### PR DESCRIPTION
**Pull request by anodium**
Due to #65, #66, #67 being light, by @jroweboy's suggestion I've merged all three months into one article to make it actually worthwhile.

Things that are still missing:

- [x] Screenshots and graphs
- [x] Banner
- [ ] Forum topic and metadata
- [x] 2017 December research
- [x] 2017 December sections

Closes #65
Closes #66
Closes #67

### CODE REVIEW

**Comment by anodium** [comment]
Review time! :postal_horn:

Let's *actually* get this reviewed, please. :tired_face:

- [ ] @bunnei
- [ ] @Subv
- [ ] @jroweboy

File: site/content/entry/citra-progress-report-2017-q4.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
++++
+date = "2018-01-21T11:24:00-04:00"
+title = "Citra Progress Report - 2017 Q4"
+tags = [ "progress-report" ]
+author = "anodium"
+forum = 0
++++
+
+As the year draws to a close, work on Citra absolutely doesn't! A lot of technical
```
Comment (bunnei) [inline site/content/entry/citra-progress-report-2017-q4.md]: I realize this is 4th quarter, but this is just confusing to users. Can you rephrase this a bit to something like "As we start the new year ... "

File: site/content/entry/citra-progress-report-2017-q4.md [hunk 2]
```diff
@@ -140,18 +140,48 @@ time it takes for the service that scheduled it would run.
...
+
+This was extremely inconvenient and user-unfriendly to most, and so [muemart](https://github.com/muemart)
+took it upon themselves to finally add support for configuring gamepads within the
+configuration menu. Now, it's as simple as clicking a button, and pressing the
+corresponsing button on your controller, to set it up.
+
 ## [shader_jit_x64_compiler: Remove ABI overhead of LG2 and EX2](https://github.com/citra-emu/citra/pull/3145) by [MerryMage](https://github.com/MerryMage)
 
--->
+The shader JIT is a subsystem in dynarmic (Citra's ARM to x86 recompiler) is the
```
Comment (jroweboy) [inline site/content/entry/citra-progress-report-2017-q4.md]: the shader jit has nothing to do with dynarmic. Its entirely its own thing in video_core

File: site/content/entry/citra-progress-report-2017-q4.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+the hood, from applets to IPC, have made Citra even more accurate and laid a
+foundation for even more goodies the next year. With many great changes all
+around, this article is going to be packed to the brim with all the new goodies
+that have come to Citra during those crimson months.
+
+But, enough faffing about! Let's get right into it:
+
+## [citra-qt : Adding fullscreen mode](https://github.com/citra-emu/citra/pull/3001) by [Styleoshin](https://github.com/Styleoshin)
+
+One of the most requested features for Citra, second only to controller support,
```
Comment (jroweboy) [inline site/content/entry/citra-progress-report-2017-q4.md]: leave out "second only to controller support" the number one request is speed ;)

File: site/content/entry/citra-progress-report-2017-q4.md [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+applications and games directly, but CXI executables can access the virtual SD
+card for DLC, updates, etc.
+
+## [Qtifw build installer](https://github.com/citra-emu/citra/pull/2966) by [j-selby](https://github.com/j-selby)
+
+Citra now has a fancy new installer and updater, thanks to [j-selby](https://github.com/j-selby)
+and [jroweboy](https://github.com/jroweboy)'s efforts! If you haven't already, we
+strongly recommend you download and install Citra through it, since it will make
+updating as easy as re-running the installer. You can check it out over [here](/download/),
+
```
Comment (jroweboy) [inline site/content/entry/citra-progress-report-2017-q4.md]: Maybe include a blurb about the "Check for update" feature in the help menu

File: site/content/entry/citra-progress-report-2017-q4.md [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+After poking around Apple's LLVM fork, [MerryMage](https://github.com/MerryMage)
+found that they had a specific flag to build binaries for Intel Haswell and later.
+Enabling it made a fat binary that contained both pre-Haswell and post-Haswell code,
+allowing Citra to take advantage of the newer CPU's instructions, without dropping
+support for older ones.
+
+## [Kernel/IPC: Add a small delay after each SyncRequest to prevent thread starvation.](https://github.com/citra-emu/citra/pull/3091) by [Subv](https://github.com/Subv)
+
+When communicating with services, titles on a real Nintendo 3DS expect to wait some
```
Comment (jroweboy) [inline site/content/entry/citra-progress-report-2017-q4.md]: This isn't correct. When a specific call is made (`SyncRequest`) the 3ds puts the calling thread to sleep until the request is done. Some games rely on this behavior as a sort of priority inversion, allowing a lower priority thread a chance to run while this thread is asleep. Because citra would respond immediately, the lower priority thread wasn't getting scheduled to run. The time that we put here was the average amount of time it took to respond, but this isn't taking into account that the delay is different for different server sessions. So in the example of reading from romfs, we likely will need a different value to delay for (more re needs to be done first)

File: site/content/entry/citra-progress-report-2017-q4.md [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+amount of time before the service replies or responds. Until now, Citra didn't implement
+this, it responded without incrementing the virtual clock. From the emulated system's
+perspective, it appeared as though services replied and reponded literally instantly
+because of this, leading to strange side-effects.
+
+For example, some buildings in *Animal Crossing: New Leaf* fail to render and
```
Comment (jroweboy) [inline site/content/entry/citra-progress-report-2017-q4.md]: add pic of ac:nl not working

File: site/content/entry/citra-progress-report-2017-q4.md [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+
+The example in this figure is taken directly from *Star Fox 3D*. When the game boots,
+it checks if any save data exists on the console's NAND, and tries to recreate it
+if it doesn't. It does this by sending a request to the service responsible for
+file access, and then waiting for a reply. Because the request is synchronous,
+the thread that made the request is put to sleep and its priority lowered. The
+game actually uses this time that would otherwise be spent waiting by having other
+threads run while the original thread was still asleep, so that by the time it
+woke up with the response, it would have other resources it needed ready to go.
+
```
Comment (jroweboy) [inline site/content/entry/citra-progress-report-2017-q4.md]: add pic of ac:nl working

File: site/content/entry/citra-progress-report-2017-q4.md [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+Because Citra's responses are instantaneous from the point-of-view of the game,
```
Comment (jroweboy) [inline site/content/entry/citra-progress-report-2017-q4.md]: because because becasue becacseu baceuaecacb euacbueabcuacbuaebcuabeubacuebauebuasaueuasecbuaebus

Comment (jroweboy) [reply site/content/entry/citra-progress-report-2017-q4.md]: (look for a different subordinating conjunction pls)

File: site/content/entry/citra-progress-report-2017-q4.md [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+the secondary thread doesn't have nearly enough time to finish its job. Because
+the first thread's priority comes back up above the secondary's once it wakes up,
+this leads to the first thread waiting on the secondary thread to finish, but it
+never gets a chance to, because of its now much lower priority. Essentially, the
+first thread waits on the secondary forever, because the secondary never gets a
+chance to actually finish what it was doing.
+
+After [B3n30](https://github.com/B3n30) did some testing with homebrew written by
```
Comment (jroweboy) [inline site/content/entry/citra-progress-report-2017-q4.md]: ds wrote the homebrew to show there was a problem. subv wrote a homebrew to get some values to test which b3n30 ran

File: site/content/entry/citra-progress-report-2017-q4.md [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+A faster algorithm, combined with not having to deal with the overhead of calling
+an external library for math functions, led to this change almost halving the
+amount of time it takes to calculate these!
+
+## Conclusion
+
+As always, [changes big or small](https://github.com/citra-emu/citra/graphs/contributors?from=2017-10-01&to=2017-12-31&type=c)
+are vital to the project. Brick by brick, Citra will eventually be an accurate enough
+emulation of the Nintendo 3DS for most, if not all, uses. Thank you everyone for
+the hard work you've poured in.
```
Comment (jroweboy) [inline site/content/entry/citra-progress-report-2017-q4.md]: add some upscaled pic here

File: site/content/entry/citra-progress-report-2017-q4.md (hunk 5 above)
Comment (jroweboy) [reply site/content/entry/citra-progress-report-2017-q4.md]: update ds wrote it https://gist.github.com/ds84182/ecdbbd25b56a29bd4e5b32a7544b8e92

File: site/content/entry/citra-progress-report-2017-q4.md (hunk 4 above)
Comment (anodium) [reply site/content/entry/citra-progress-report-2017-q4.md]: I didn't realize there was an update feature from within Citra, since every build I've checked (including the one I compiled from `master` a few hours ago) doesn't have a "Check for Update" button in the Help menu. Is it only enabled with a CMake switch?

**Comment by jroweboy** [comment]
you have to use the installer to get the button.

File: site/content/entry/citra-progress-report-2017-q4.md (hunk 4 above)
Comment (FearlessTobi) [reply site/content/entry/citra-progress-report-2017-q4.md]: @anodium I believe only builds from the installer have the "Check for Updates" button.
